astropy/utils/data.py

```diff
@@ -244,16 +246,49 @@ def get_readable_fileobj(name_or_obj, encoding=None, cache=False,
             delete_fds.append(tmp)
             fileobj_new = bz2.BZ2File(tmp.name, mode='rb')
             fileobj_new.read(1)  # need to check that the file is really bzip2
-        except IOError:  # invalid bzip2 file
+        except (IOError) as e:  # invalid bzip2 file
             fileobj.seek(0)
             fileobj_new.close()
+            # raise
         else:
             fileobj_new.seek(0)
             close_fds.append(fileobj_new)
             fileobj = fileobj_new
+    elif signature[:3] == b'\xfd7z':  # xz
+        try:
+            # for Python < 3.3 try backports.lzma; pyliblzma installs as lzma,
+            # but does not support TextIOWrapper
+            if sys.version_info >= (3,3,0):
+                import lzma
+                fileobj_new = lzma.LZMAFile(fileobj, mode='rb')
```

Comment:
Pretty sure I tested for open files both with `lzma` available and not in Python 2.7, so it seemed like the `delete_fds.append(tmp)` already took care of that. In Python 3 it would not turn up, as its `lzma` module (if compiled) does not need the tmpfile workaround. There could perhaps be pathological cases of Python 3 with `lzma` support initially not compiled in, but `pyliblzma` installed later... Anyway will test it again as well.